feat: impl bootstrap cache

RolandSherwin opened the pull request:
- Introduces `ant-bootstrap` crate which replaces the `ant-peers-acquisition` crate.
- Every node now by default writes to a single `bootstrap_cache_<network_id>.json` file in the data dir. It contains a map of PeerId to their addresses. We track the success/failure of each address.
- These files are synced on periodic interval. Nodes don't overwrite, but instead `sync` the values.
- `network-contacts` feature flag is now removed, and now we use a `--testnet` flag to denote if we should connect to the mainnet or to a testnet.

dirvine committed:
feat(networking): add bootstrap cache for peer discovery

Add persistent bootstrap cache to maintain a list of previously known peers,
improving network bootstrapping efficiency and reducing cold-start times.

Enhance the bootstrap cache implementation with robust corruption detection
and recovery mechanisms. This change ensures system resilience when the
cache file becomes corrupted or invalid.

Key changes:
* Add explicit cache corruption detection and error reporting
* Implement cache rebuilding from in-memory peers or endpoints
* Use atomic file operations to prevent corruption during writes
* Improve error handling with specific error variants
* Add comprehensive test suite for corruption scenarios

The system now handles corruption by:
1. Detecting invalid/corrupted JSON data during cache reads
2. Attempting recovery using in-memory peers if available
3. Falling back to endpoint discovery if needed
4. Using atomic operations for safe cache updates

Testing:
* Add tests for various corruption scenarios
* Add concurrent access tests
* Add file operation tests
* Verify endpoint fallback behavior

- Add smarter JSON format detection by checking content structure
- Improve error handling with specific InvalidResponse variant
- Reduce unnecessary warnings by only logging invalid multiaddrs
- Simplify parsing logic to handle both JSON and plain text formats
- Add better error context for failed parsing attempts

All tests passing, including JSON endpoint and plain text format tests.

feat(bootstrap_cache): implement circuit breaker with exponential backoff

- Add CircuitBreakerConfig with configurable parameters for failures and timeouts
- Implement circuit breaker states (closed, open, half-open) with state transitions
- Add exponential backoff for failed request retries
- Update InitialPeerDiscovery to support custom circuit breaker configuration
- Add comprehensive test suite with shorter timeouts for faster testing

This change improves system resilience by preventing cascading failures and
reducing load on failing endpoints through intelligent retry mechanisms.

dirvine committed:
refactor(bootstrap_cache): improve peer source handling and test network isolation

* Refactor CacheStore::from_args to handle peer sources more consistently
* Ensure test network mode is properly isolated from cache system
* Fix default behavior to use URL endpoint when no peers provided
* Add proper handling for local and first node modes
* Prevent cache operations when in test network mode

This change ensures that:
- Test network peers are isolated from cache operations
- Default behavior (no args) correctly uses URL endpoints
- Local and first node modes return empty stores
- Explicit peers take precedence over default behavior
- Cache operations only occur in non-test network mode

The changes make the peer source handling more predictable and maintain
proper isolation between different network modes (test, local, default).

dirvine committed:
chore: update readme

dirvine committed:
fix(bootstrap_cache): improve test isolation and env var handling

* Fix test_safe_peers_env to verify env var peer inclusion
  - Assert presence of env var peer in total peer set
  - Remove incorrect assertion of exact peer count

* Fix test_network_contacts_fallback isolation
  - Enable test_network mode to prevent interference from cache/endpoints
  - Verify exact peer count from mock server

* Improve from_args implementation
  - Add environment variable peer handling before other sources
  - Use empty cache path in test network mode
  - Prevent cache file operations in test network mode

These changes ensure proper test isolation and correct handling of peers
from different sources (env vars, args, cache, endpoints) across different
modes (normal, test network, local).

RolandSherwin committed:
fix(bootstrap_cache): remove unused code and prep it for integration

- prep the cache_store to write to disk on periodic interval rather than
  on every op
- use the default config dir that is being used through out the codebase
- use simple retries for network GETs rather than using complex backoff

RolandSherwin committed:
fix(bootstrap): remove rwlock from the store

RolandSherwin committed:
feat(bootstrap): wrap the counts when reaching the max bounds

RolandSherwin committed:
fix(bootstrap): couple more tiny fixes

RolandSherwin committed:
feat(bootstrap): store multiple multiaddr per peer

RolandSherwin committed:
feat(bootstrap): isolate code into their own modules based on their purpose

b-zee commented on ant-peers-acquisition/Cargo.toml:
```diff
@@ -10,6 +10,7 @@ repository = "https://github.com/maidsafe/autonomi"
 version = "0.5.7"
 
 [features]
+default = ["network-contacts"]
```
Maybe we should integrate `ant-peers-acquisition` into `ant-bootstrap`? So we have just one crate related to bootstrap contacts.

RolandSherwin replied:
Yes! That is the plan, we'd use the `ant-bootstrap::PeersArgs` instead of the other one.

b-zee replied:
Cool!

b-zee commented on Justfile:
```diff
@@ -68,15 +68,15 @@ build-release-artifacts arch nightly="false":
     cross build --release --target $arch --bin nat-detection $nightly_feature
     cross build --release --target $arch --bin node-launchpad $nightly_feature
-    cross build --release --features network-contacts,websockets --target $arch --bin autonomi $nightly_feature
+    cross build --release --features network-contacts,websockets --target $arch --bin ant $nightly_feature
```
Could we merge these changes in a separate PR?

RolandSherwin replied:
Sorry I think that was a previous git commit that somehow snuck in here, it is updated now.

b-zee commented on .gitignore:
```diff
@@ -36,8 +36,7 @@ sn_node_manager/.vagrant
 *.pyc
 *.swp
 
+/vendor/
```
What is this for?

RolandSherwin replied:
No idea, it was from David's commit.

jacderida commented on ant-peers-acquisition/Cargo.toml:
```diff
@@ -10,6 +10,7 @@ repository = "https://github.com/maidsafe/autonomi"
 version = "0.5.7"
 
 [features]
+default = ["network-contacts"]
```
Could we just lose this as a 'feature'? I think that was very excessive anyway.

RolandSherwin replied:
Yes! We'd remove this.
The `network-contacts` just indicates if we want to fetch the `mainnet` contacts or not.

This can be replaced by a simple `--testnet` flag.

RolandSherwin committed:
feat(bootstrap): impl bootstrap cache into the codebase

RolandSherwin committed:
feat: remove ant-peers-acquisition and use ant-bootstrap instead

- This also removes the `network-contact` feature flag.
- The flag was used to indicate if we should connect to the mainnet or
  the testnet, which can easily be done with PeersArgs::testnet flag

RolandSherwin committed:
fix(bootstrap): use env tempdir for atomic write

RolandSherwin committed:
fix(bootstrap): make it wasm compatible

RolandSherwin committed:
fix(bootstrap): use atomic write crate and remove locks

maqi commented on ant-networking/src/driver.rs:
```diff
@@ -1005,6 +1025,39 @@ impl SwarmDriver {
+                        continue;
+                    };
+
+                    if let Err(err) = bootstrap_cache.sync_and_save_to_disk(true).await {
```
this will be risky as will slow down the `swarm_driver` handling, and may even result in `dead-lock`, if running multiple nodes on one machine.
`swarm_driver` shall only handle small/instant tasks. any `async` functions (potentially could be long or blocked by others) shall be passed to a spwaned task to handle, or even fired as an event to higher layer (network) to be handled.

RolandSherwin committed:
feat(ci): enable bootstrap tests

maqi commented:
the memcheck CI run has a step of `Start a node instance to be restarted` , which currently using : 
```code
./target/release/antnode --root-dir $RESTART_TEST_NODE_DATA_PATH --log-output-dest $RESTART_TEST_NODE_DATA_PATH --local --rewards-address "0x03B770D9cD32077cC0bF330c13C114a87643B124" &
```
and relies on the `ANT_PEERS` env to be set.

It will be the best to refactor it to use `bootstrap cache` shared_file to startup with, so that the work can be verified by CI directly.

RolandSherwin committed:
feat(bootstrap): rework the api to not hold persistant state

RolandSherwin committed:
chore(bootstrap): remove components related to serving the json

RolandSherwin committed:
feat(bootstrap): write bootstrap cache from the clients

RolandSherwin committed:
chore(antctl): use PeersArg::local instead of a separate arg

maqi commented on ant-networking/src/driver.rs:
```diff
@@ -1005,6 +1025,58 @@ impl SwarmDriver {
+                    // scale up the interval until we reach the max
+                    let new_duration = Duration::from_secs(
+                        std::cmp::min(
+                         current_interval.period().as_secs() * bootstrap_cache.config().cache_save_scaling_factor,
```
just to ensure that an overflow multiple won't cause a panic

maqi commented on ant-networking/src/driver.rs:
```diff
@@ -1005,6 +1025,58 @@ impl SwarmDriver {
+                    let config = bootstrap_cache.config().clone();
+                    let mut old_cache = bootstrap_cache.clone();
+
+                    let new = match BootstrapCacheStore::new(config) {
```
doesn't need to have a new one ?
just cotinue with the existing one shall be ok?

anyway, this depends on how the BootstrapCache handles `merge`.

RolandSherwin replied:
The existing one will contain the tracked peers and will only be cleared after calling `flush`. That is spawned as a separate task, so we cannot clone it as well. So need a new one here.

maqi commented on ant-networking/src/event/swarm.rs:
```diff
@@ -506,6 +515,15 @@ impl SwarmDriver {
+                // Just track failures during outgoing connection with `failed_peer_id` inside the bootstrap cache.
+                // OutgoingConnectionError without peer_id can happen when dialing multiple addresses of a peer.
+                // And similarly IncomingConnectionError can happen when a peer has multiple transports/listen addrs.
+                if let (Some((_, failed_addr, _)), Some(bootstrap_cache)) =
```
I think this block of code can be moved into the below block of `if should_clean_peer {` ?
where a peer is for sure hits a serious issue and got removed from the RT ?

maqi commented on ant-node/src/node.rs:
```diff
@@ -160,11 +177,25 @@ impl NodeBuilder {
             None
         };
 
+        if !self.initial_peers.is_empty() && self.bootstrap_cache.is_some() {
```
this block (180 - 191) of code can be down in : 
```rust
let initial_peers = match (self.initial_peers.is_empty(), self.bootstrap_cache) {
    (true, Some(_)) => return Err(Error::InitialPeersAndBootstrapCacheSet),
    (true, None) => self.initial_peers.clone(),
    (false, Some(cache)) => cache.get_sorted_addrs().cloned().collect(),
    (false, None) => vec![],
};
```
Which will be more readable ?

also, shall `initial_peers` always be preferred, i.e. it shall be `(true, _) => self.initial_peers.clone(),` ?

RolandSherwin replied:
Ahh.. we cannot get the initial bootstrap peers from the cache store anylonger as it won't be initialized anymore. 
The loading of cache happens at a higher layer and it is fed into `initial_peers`. I will modify according to this, thanks for the catch!

maqi commented on ant-bootstrap/src/lib.rs:
```diff
@@ -0,0 +1,254 @@
+        self.success_count = self.success_count.saturating_add(other.success_count);
+        self.failure_count = self.failure_count.saturating_add(other.failure_count);
+
+        // if at max value, reset to 0
```
shall be `capped`, not `reset` ?
i.e. no need to take this reset step

RolandSherwin replied:
This won't be an issue in the real world, but for some reason if we were to reached the max for `success`, then only the `failures` will be tracked until it catches up with the `success`, making this peer useless as it will not be reliable (success = failure = max)

So just wrap around to 0.

RolandSherwin committed:
chore: update based on comments
